apps/server/src/modules/learnroom/service/course-sync.service.ts

```diff
@@ -0,0 +1,76 @@
+import { Group, GroupUser } from '@modules/group';
+import { RoleService } from '@modules/role';
+import { Inject, Injectable } from '@nestjs/common';
+import { RoleName } from '@shared/domain/interface';
+import { EntityId } from '@shared/domain/types';
+import {
+	Course,
+	COURSE_REPO,
+	CourseAlreadySynchronizedLoggableException,
+	CourseNotSynchronizedLoggableException,
+	CourseRepo,
+} from '../domain';
+
+@Injectable()
+export class CourseSyncService {
+	constructor(
+		@Inject(COURSE_REPO) private readonly courseRepo: CourseRepo,
+		private readonly roleService: RoleService
+	) {}
+
+	public async startSynchronization(course: Course, group: Group): Promise<void> {
+		if (course.syncedWithGroup) {
+			throw new CourseAlreadySynchronizedLoggableException(course.id);
+		}
+
+		await this.synchronize(course, group);
+	}
+
+	public async stopSynchronization(course: Course): Promise<void> {
+		if (!course.syncedWithGroup) {
+			throw new CourseNotSynchronizedLoggableException(course.id);
+		}
+
+		course.syncedWithGroup = undefined;
+
+		await this.courseRepo.save(course);
+	}
+
+	public async synchronizeCourseWithGroup(newGroup: Group, oldGroup?: Group): Promise<void> {
+		const courses: Course[] = await this.courseRepo.findBySyncedGroup(newGroup);
+
+		if (courses.length) {
+			await Promise.all(
+				courses.map(async (course) => {
+					await this.synchronize(course, newGroup, oldGroup);
+				})
+			);
+		}
+	}
+
+	private async synchronize(course: Course, group: Group, oldGroup?: Group): Promise<void> {
+		course.syncedWithGroup = group.id;
+		if (!oldGroup || oldGroup.name === course.name) {
```

Comment:
Since the name should not change, when the sync is started, the `!oldGroup` should not be a condition anymore. I think this condition was never a thing, since you cannot have a synchronized course with a group that was freshly provisioned.

Reply:
I misunderstood this comment, I reverted back some of the changes